src/harness/reference_models/pre_iap_filtering/fss_purge.py

```diff
@@ -0,0 +1,299 @@
+                                               height_agl=fss_height_agl,
+                                               max_gain_dbi=fss_max_gain_dbi,
+                                               pointing_azimuth=fss_azimuth,
+                                               pointing_elevation=fss_pointing_elevation)
+  return fss_entity
+
+def generatePurgeListForFssPoint(cbsds, fss_entity):
+  """ Implementation of algorithm described in R2-SGN-29.
+
+  The list of grants to be purged based on the interference caused by their
+  out of band emission are identified and returned.
+  Args:
+    cbsds: List of CbsdData objects that are in the neighborhood to an FSS.
+    fss_entity: The FssProtectionPoint tuple of the FSS for which the purge is to be performed.
+  Returns:
+    List of multiple key-value pairs with keys "grantData", "cbsd" and values as grants to be
+    purged and corresponding cbsd data object respectively.
+  """
+  # List of grants to be purged for the fss_entity.
+  final_purge_list = []
+
+  # List of dictionaries containing the grant_info based on the highest grant from FSS OOBE calculation
+  grants_cbsds_info_for_oobe_calculation = getGrantInfoListForFssOobeCalculation(cbsds)
+
+  # Calculate the threshold value in dBm for the OOBE interference
+  oobe_threshold_value = -129 + 10 * (np.log10(REF_BW/interference.ONE_MHZ) - FSS_OOBE_MARGIN)
+
+  while True:
+
+    # Calculate the OOBE interference value for each grant.
+    calculateOobeInterference(grants_cbsds_info_for_oobe_calculation, fss_entity)
+
+    # Sort the grants based on the calculated interference value.
+    sorted_grants_cbsds_info = sorted(grants_cbsds_info_for_oobe_calculation, key=lambda grant: grant['oobe_interference'])
+    # Find the largest index such that aggregation of the interferences do not cross
+    # the OOBE threshold.
+    oobe_interferences = [interference.dbToLinear(grant['oobe_interference']) for grant in sorted_grants_cbsds_info]
+    cumulated_interference = np.cumsum(oobe_interferences)
+    index = np.searchsorted(cumulated_interference, interference.dbToLinear(oobe_threshold_value))
+    grants_to_purge = sorted_grants_cbsds_info[index:]
+    grants_cbsds_info_for_oobe_calculation = sorted_grants_cbsds_info[:index]
+
+    if grants_to_purge:
+      for purged_grant in grants_to_purge:
+        remaining_grants = []
+        for cbsd_grant in purged_grant['cbsd']['grants']:
+          # Find the grants matching mcbsd value and add it to the purge list
+          if (purged_grant['mcbsd']) == getMcbsdValue(cbsd_grant):
+            final_purge_list.append(grants_cbsds_namedtuple(cbsd_grant, purged_grant['cbsd']))
+          else:
+            # Find the grants not matching the mcbsd values and not present in the purge list
+            # and add it to the remaining grants list
+            for final_grant in final_purge_list:
+              if cbsd_grant is final_grant.grant:
+                break
+            else:
```

Comment:
There isn't a need to check the ```final_purge_list```, ignore the grant unless the grants mcbsd is lower than the mcbsd in purged_grant.

```python
if purged_grant['mcbsd'] == getMcbsdValue(cbsd_grant):
...
elif purged_grant['mcbsd'] > getMcbsdValue(cbsd_grant):
  remaining_grants.append(cbsd_grant)
```

Reply:
updated   the code as comment suggested